core/sr-primitives/src/weights.rs

```diff
@@ -0,0 +1,85 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Primitives for transaction weighting.
+//!
+//! Each dispatch function withing `decl_module!` can now have an optional
+//! `#[weight = $x]` attribute. $x can be any object that implements the
+//! [`WeighableTransaction`] trait. By default, All transactions are annotated by
+//! `#[weight = TransactionWeight::default()]`.
+//!
+//! Note that the decl_module macro _cannot_ enforce this and will simply fail
+//! if an invalid struct is passed in.
+//!
+//! Note that [`WeighableCall`] and [`WeighableTransaction`] are more or less similar.
+//! The distinction is because one serves to pass the weight from the the
+//! dispatchable function's attribute to the call enum ([`WeighableTransaction`]) and the
+//! other to pass the final weight from call enum to the executive module
+//! ([`WeighableCall`]).
+
+/// The final type that each `#[weight = $x:expr]`'s
+/// expression must evaluate to. Consists of `(base_weight, byte_weight)`.
+pub type Weight = (u32, u32);
+
+// TODO #2431 this value is now set in a way that the weight in the transaction
+// is just the size of input. Must be updated based based on benchmarks. Same for
+// maximum_weight and ::max() variant of TransactionWeight.
+/// The default weight as literal value.
+pub const DEFAULT_WEIGHT: Weight = (0, 1);
+
+/// A `Call` enum that can be weighted using the custom weight attribute of the
+/// its dispatchable functions. Is implemented by default in the `decl_module!`.
+pub trait WeighableCall {
+	/// Return the weight of this call.
+	fn weight(&self) -> Weight;
+}
+
+/// a _dispatchable_ function (anything inside `decl_module! {}`) that can be weighted.
+/// A type implementing this trait can _optionally_ be passed to the as
+/// `#[weight = X]`. Otherwise, default implementation will be used.
+pub trait WeighableTransaction {
```

Comment:
I don't feel this trait is very useful. The `WeighableCall` interfaces is limitting this stuff to linear cost of  `base` and `bytes` , I think that it might be worth allowing users to specify different cost calculations that are based on the call and it's encoded len, which is not possible currently.

Reply:
I had the idea of merging them but fyi this is the explanation:
https://github.com/paritytech/substrate/pull/2799/files#diff-74eb2b089bc101850d8e1b995695c0d6R27

Reply:
I still think it's unnecessary. `WeighableCall` and `WeighableTransaction` look almost identical - they just have different method names and in transaciton `self` is consumed (for no apparent reason).

I'd unify this into `Weighable` and implement for multiple things if needed.